### ISSUE TITLE
How do you recommend to handle xNode version conflicts?

### ISSUE THREAD

**Issue by OscarAbraham** [Siccity/xNode#211]
Hi, xNode is awesome; thank you. I was looking at the repository and I've found that it's API is susceptible to change. So, if a project uses two package that use a specific version of xNode, it can cause conflicts. It seems probable that any conflict resolution that keeps only one version of xNode, would make some of the packages not work. Am I right?

Is there any recommendation to make a library or a package that uses xNode plays well with other libraries or packages that use Xnode? So far, the only thing I can think of is to edit the xNode scripts in my own package so that the use another namespace, it doesn't seem ideal because it makes it harder to update them. Is that currently the only way of handling this problem? Thank you for your time.

**Comment by StefanoCecere** [comment]
i, too, have found several examples that use older version of xNode.. and aren't compatible with current master :(

**Comment by Siccity** [comment]
I try to keep API changes to a minimum, but sometimes it happens anyway.

My best advice would be to try to update the package that is using an older version. Typically the changes required are not very complex (renaming "instance ports" to "dynamic ports" or using GetNoodleGradient instead of GetNoodleColor etc.)

Ideally you keep only one, the most updated version of xNode, in your project.

Can you give some examples of incompatible projects?

**Comment by OscarAbraham** [comment]
Thank you very much, really. Uhm, I haven't found a use case where it's been a problem for me directly. The issue occurred to me because I made a tool to control monobehaviours with a State Machine. It has been helping me a lot, so I thought I'd make it a little more presentable and then share it as a package on Github. What worries me a little about that is that I don't have a control of how other users or maintainers update the projects to use the most updated version of xNode. I've seen that most of the packages in Github or the Asset Store that use xNode just embed it on their own folder.

I imagine this most be an issue that it's common to many open source projects for Unity. I'm sorry for rising such a noob question. Most of my experience with packages comes from the npm land, where the package manager just injects dependencies you need where you need them, in a version that works for you. I suppose that's not even practical to do with C# projects. Most of the open source packages that I've worked with on Unity, until now, are not destined to build tools on top of them. The few packages I've tried that are tools to make tools, are either very mature and consolidated, like some serializers, so their API doesn't change, or very small, like property drawers, so they are easier to maintain as a forked version under their own namespace.

**Comment by Siccity** [comment]
Sorry for the late response. Have you heard about git submodules? If you add xNode as a submodule to your project, git will point to a specific commit of xNode, meaning people who download your project will get that exact version that you know works. From then on, you or your users have the option to update xNode independently. The downside to this, is that many people find submodules confusing, so they can't take full advantage of it.

**Comment by gekidoslair** [comment]
If Xnode is wrapped with an assembly definition, you "might/should" be able to have multiple in a project, as long as each dependent library references the proper assembly

**Comment by OscarAbraham** [comment]
I'm really sorry. I forgot this issue I had raised. That's just what I ended up doing, well, at least for a while. There is still some potential conflict with the preferences panel and using of node and graph classes outside the assembly definition, but those are issues that can be worked around. 

I ended up putting it all inside another namespace though, because what I'm doing has a lot more to do with nodes being connected between themselves, than ports being some sort of surrogates for that connection. It was a lot easier to add all the code I needed to represent that kind of connection directly to the library than using subclasses. Then, it was easier to think about by removing all the things I didn't need, like ports and so. The code doesn't really look a lot like xNode now, so I had to give up on being able to update it from this repository anyway, which kind of solved the whole version conflict concern for me.

But, dear future reader, it still is a good idea to use assembly definitions if you are worried about this kind of thing. Basically, go to the main asmdef in the xnode folder, and turn off Auto Referenced. Do that for the editor asmdef too. Then, go to the asmdefs in your main project that need to know about xNode (including the editor xNode asmdef), and add the relevant definitions to the Assembly Definition references lists.  This way, only code that needs to know about xNode, knows about it. You might still want to handle the problems mentioned in the first paragraph though.

Any way, I'm closing the issue; I think all the possible solutions have been mentioned. Thank you all for your time. This library has helped me a lot to understand how Unity's editor works and I'm still using a good portion of it, so thank you very much.
